grin/test/Test/EndToEnd.hs

```diff
@@ -0,0 +1,224 @@
+  actionWith $ \() -> catch
+    (do let fileArgs = case input of
+                        Binary fp  -> [fp, "--load-binary"]
+                        Textual fp -> [fp]
+        let evalArgs = ["--output-dir=.end-to-end-test", "--quiet", "--eval"]
+        (grinOut, ()) <- redirectStdout $ mainWithArgs $ fileArgs ++ evalArgs
+        removeDirectoryRecursive ".end-to-end-test"
+        let compArgs =
+              [ "--output-dir=.end-to-end-test"
+              , "--quiet"
+              , "--save-binary-intermed"
+              , "--optimize"
+              , "--save-elf=end-to-end-test.bin"
+              , "--runtime-c-path=./grin/test-runtime/runtime.c"
+              , "--primops-c-path=./grin/prim_ops.c"
+              ]
+        mainWithArgs $ fileArgs ++ compArgs
+        let runTest = (shell "./end-to-end-test.bin")
+                      { std_in = NoStream, std_out = CreatePipe, std_err = CreatePipe }
+        (mIn, Just out, Just err, runTestPh) <- createProcess_ "./end-to-end-test.bin" runTest
+        runTestExitCode <- waitForProcess runTestPh
+        doesFileExist "./end-to-end-test.bin" >>= flip when (removeFile "./end-to-end-test.bin")
+        testOut <- hGetContents out
+        if (grinOut == fromString testOut)
+          then writeIORef result $ Result "" Success
+          else do
+            writeIORef result $ Result "" $ Failure Nothing $ Reason "End-to-end test started bisecting but it did not finish."
+            res <- bisect ".end-to-end-test" grinOut
+            writeIORef result res
+    )
+    (writeIORef result . Result "" . Failure Nothing . Error Nothing)
+  readIORef result
+
+loopM :: (Monad m) => (a -> m (Either a b)) -> a -> m b
+loopM n a0 = n a0 >>= \case
+  Left a  -> loopM n a
+  Right b -> pure b
+
+class Monad m => BisectM m where
+  createFileMap :: FilePath -> m (Map.Map Int FilePath)
+  runTest       :: FilePath -> ByteString -> m Bool
+
+instance BisectM IO where
+  createFileMap directory =
+      fmap (cFileMap . filter isGrinFile) $ listDirectory directory
+    where
+      noOfDigits = 3
+      isGrinFile name = (all isDigit (take noOfDigits name)) && ".binary" `isSuffixOf` name
+      cFileMap files = Map.fromList $
+        [ (itr, directory </> name)
+        | name <- files
+        , let itr = read @Int (take noOfDigits name)
+        ]
+
+  runTest file exp = do
+    let compArgs =
+          [ file
+          , "--quiet"
+          , "--load-binary"
+          , "--eval"
+          ]
+    (grinOut, ()) <- redirectStdout $ mainWithArgs compArgs
+    pure $ grinOut == exp
+
+bisect :: (BisectM m) => FilePath -> ByteString -> m Result
+bisect directory expected = do
+  fileMap <- createFileMap directory
+  let (mn, mx) = findRange fileMap
+  let fileToTest x = fromMaybe (error $ show mn) $ Map.lookup x fileMap
+  tn <- runTest (fileToTest mn) expected
+  tx <- runTest (fileToTest mx) expected
+  loopM (go fileMap) ((mn,tn), (mx,tx))
+  where
+    go fm ((mn,tn), (mx, tx))
```

Comment:
`go` is a no-go. Please use a more descriptive name or add some Haddocs. In either case, please also add a type sig.

Reply:
`loop go` for simpler problems is a practice. But you are right this function grew beyond being simple. Changing...